design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+the following additions:
+ - `void setProgressBar(ProgressBar progressBar)`
+ - `void setText(String text)` (inherited from `Label`)
+ - `void render()` (we can probably add this to `Label`)
+ 
+Under `--continuous` build, the progress bar and multi status text area are reused.
+
+### Test Coverage
+* By default, logs should be streamed with plain output and not throttled when not attached to a Console
+* Color output is output when requested through via the tooling API via `LongRunningOperation.setColorOutput(true)`
+* OutputEvents tooling to API is unaffected except for additional BuildOperations now published
+* Logs should be streamed with plain output when attached Terminal lacks of color or cursor support
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, PowerShell and common Linux Terminals
+
+## Story: Display work-in-progress through ProgressBar
+This adds a visual indicator of what proportion of the build would be complete if all in-progress work is completed.
+This requires the use of color to distinguish between different regions.
+
+This turns this:
+`###############>                          > 40% Building`
+
+into this:
+`#####green#####>##yellow##>#####black#####> 40% Building`
+
+### Implementation
+A concept of `DiscreteProgress` wraps current, in-progress, and total fields, each Longs.
+
+`ProgressBar` would have another operation:
+ - `void setRegions(List<ProgressBarRegion> regions)`
+
+A `ProgressBarRegion` has operations:
+ - `void setPrefix(String prefix)`
+ - `void setSuffix(String suffix)`
+ - `void render(Integer width)`
+ 
+`BuildProgressBarListener` would keep track of work starting through `projectsEvaluated(Gradle)`
+and `beforeExecute(Task)` and forward that information to `ProgressLabel` 
+
+### Open Issues
+* Alternative implementation of text-based spinner (e.g. "┘└┌┐" or "|/-\\")
+
+## Story: Display intra-operation progress
+When we know in advance the number of things to be processed, we display progress in [Complete / Total] format. 
+Provide APIs to optionally add # complete and # of all items.
+
+For example:
+```
+[23 / 65] Configuring projects
+[56 / 245] Resolving dependencies for configuration 'compileJava'
+```
+
+### Implementation
+`BuildOperationContext` gains a method to report a `DiscreteProgress`.
+While executing a `BuildOperation`, progress can be optionally reported
+by a ProgressLogger that is routed to the a Console renderer.
+
+`SimpleProgressFormatter#getProgress()` format is changed to wrap `current` and
+`total` in square brackets.
+
+`BuildProgressBarListener` implements `DependencyResolutionListener` and
+forwards dependency resolution events to `BuildProgressBarLogger`.
+
+### Test Cases
+* Implementations are thread-safe in the face of multiple threads updating status
+
+### Open Issues
+* Can we discover the number of tests to run or classes to compile ahead of time?
+* We'll need a public API to give plugins access to log intra-operation progress.
```

Comment:
Great, I'm very happy for users to have that. IIRC we still intend to provide that for Gradle 4.0, correct?